### PR TITLE
Utilize v2 targetprocess API for GETs

### PR DESCRIPTION
**Pull request by lucasreed**
Changes:
- BREAKING: removed Include() filter as it is not available in the v2
targetprocess API
- BREAKING: removed AssignedTeams from UserStory default struct (custom
structs can still include any fields a user would like)
- BREAKING: changed method signature of client.GetUserStories() to
include a boolean to determine if it should automatically page through
results
    - This was important for using the First() filter in this method. If
the method automatically paged with the First() filter function it would
return ALL results 1 UserStory at a time which is a lot of API calls and
woulud take a really long time.
- added ResponsibleTeam to the UserStory default struct
- added Select() and Result() filters
- added tests for each filter

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -36,8 +36,11 @@ const (
 // Client is the API client for Targetprocess. Create this using NewClient.
 // This can also be constructed manually but it isn't recommended.
 type Client struct {
-	// BaseURL is the base URL for API requests.
-	BaseURL *url.URL
+	// baseURL is the base URL for v1 API requests.
+	baseURL *url.URL
```
Comment (sudermanjr) [inline client.go:40]: I see why this is unexported, but can you envision any scenario in which the consumer might want to access this? Possibly to make custom http requests to the URL that aren't covered by this library?

File: client.go [hunk 2]
```diff
@@ -74,16 +77,22 @@ func NewClient(account, token string) *Client {
...
 	c.Timeout = 15 * time.Second
 	baseURLString := fmt.Sprintf("https://%s.tpondemand.com/api/v1/", account)
+	baseURLReadOnlyString := fmt.Sprintf("https://%s.tpondemand.com/api/v2/", account)
 	baseURL, err := url.Parse(baseURLString)
 	if err != nil {
 		panic(err)
 	}
+	baseURLReadOnly, err := url.Parse(baseURLReadOnlyString)
+	if err != nil {
+		panic(err)
```
Comment (sudermanjr) [inline client.go:87]: I know this is what we had before, but is `panic` correct here? Should we be returning an error instead?

File: client.go (hunk 1 above)
Comment (sudermanjr) [reply client.go:40]: I guess that's covered by the generic `GET` method, so probably not an issue.

File: client.go [hunk 3]
```diff
@@ -146,40 +161,41 @@ func (c *Client) Post(out interface{}, entityType string, values url.Values, bod
...
 	if err != nil {
-		return errors.Wrapf(err, "HTTP request failure on %s/%s", c.BaseURL, urlPath)
+		return errors.Wrapf(err, "HTTP request failure on %s", noParameterURL)
 	}
 
 	// Empty the body and close it to reuse the Transport
 	defer func() {
-		io.Copy(ioutil.Discard, resp.Body) // nolint:golint,errcheck
-		resp.Body.Close()
+		_, _ = io.Copy(ioutil.Discard, resp.Body) // nolint:golint,errcheck
```
Comment (sudermanjr) [inline client.go:197]: Is the linter comment necessary? I thought `_, _ :=` would take care of the linter.

File: filters.go [hunk 4]
```diff
@@ -18,54 +18,78 @@ import (
...
-				values.Set("where", fmt.Sprintf("(%s)", query))
-			}
+		values.Set("take", "1")
+		return values, nil
+	}
+}
+
+// MaxPerPage is how many results we will allow to return per page. Default enforced by the API is 25.
+// If a negative number is passed in it will be converted to a positive number. In the TP API, a negative
+// number has the same return as a positive, so we'll
```
Comment (sudermanjr) [inline filters.go:41]: so we'll ......

File: filters.go [hunk 5]
```diff
@@ -18,54 +18,78 @@ import (
...
 
-// First is a QueryFilter that will only return a single
-// Entity. It returns the first one encountered.
-func First() QueryFilter {
+// Result is a QueryFilter that represents the `result` parameter
+// in a url query. It is used to do custom calculations over the
+// entire result set, such as getting the average Effort value
+// across multiple items.
+//
+// It is important to not that this changes the output json and
```
Comment (sudermanjr) [inline filters.go:57]:
```suggestion
// It is important to note that this changes the output json and
```

File: general.go [hunk 6]
```diff
@@ -30,3 +29,8 @@ type TeamAssignment struct {
 	EndDate   DateTime `json:",omitempty"`
 	Team      *Team    `json:",omitempty"`
 }
+
+// UserAssignment has it's own unique Id and also includes a reference to a user, which also has an Id
```
Comment (sudermanjr) [inline general.go:33]:
```suggestion
// UserAssignment has its own unique Id and also includes a reference to a user, which also has an Id
```

**Review by sudermanjr — CHANGES_REQUESTED** [review]
Looks good to me. Untested, but we can test after merge and then do a major release for the breaking changes.

I left some questions/suggestions inline as well.

File: client.go (hunk 2 above)
Comment (lucasreed) [reply client.go:87]: I think you're correct about this. I'll fix it up.

File: client.go (hunk 3 above)
Comment (lucasreed) [reply client.go:197]: Ah you know what I didn't think about that when I changed it. My IDE still complained so I did what it told me, but now I think that comment should have stopped the IDE from complaining. I'll revisit.

**Comment by lucasreed** [comment]
I fixed all the initial review comments. Leaving in draft for now because I will add more testing.

**Comment by lucasreed** [comment]
I think this is actually good enough and a separate PR for more tests will be easier to parse so there are fewer changes in this already fairly large PR.

**Pull request merged**